src/control/grouphandle.cpp

```diff
@@ -63,19 +122,56 @@ int resetAllGroupHandles() {
     tmpGroupHandlesByName.swap(allGroupHandlesByName);
     allGroupHandlesFrozen.storeRelease(0);
     locked.unlock();
-    for (const auto groupHandle : qAsConst(tmpGroupHandlesByName)) {
+    for (const auto& groupHandle : qAsConst(tmpGroupHandlesByName)) {
         qInfo() << "Deleting" << groupHandle;
-        delete groupHandle;
+        delete groupHandle.m_pImpl;
     }
     return tmpGroupHandlesByName.size();
 }
 
-namespace mixxx {
-namespace grouphandle_private {
+GroupHandle::GroupHandle(mixxx::grouphandle_private::GroupHandleImpl* pImpl)
+        : m_pImpl(pImpl) {
+    DEBUG_ASSERT(!m_pImpl || m_pImpl->valid());
+}
 
-QDebug operator<<(QDebug dbg, const mixxx::grouphandle_private::Descriptor& arg) {
-    return dbg << "GroupHandle{" << arg.m_index << arg.m_name << "}";
+qhash_seed_t GroupHandle::qHashImpl(qhash_seed_t seed) const {
+    if (m_pImpl) {
+        return qHash(*m_pImpl, seed);
+    } else {
+        return qHash(mixxx::grouphandle_private::GroupHandleImpl{}, seed);
+    }
 }
 
-} // namespace grouphandle_private
-} // namespace mixxx
+bool operator<(GroupHandle lhs, GroupHandle rhs) {
+    if (!rhs.m_pImpl) {
+        return false;
+    }
+    if (!lhs.m_pImpl) {
+        return true;
+    }
+    return *lhs.m_pImpl < *rhs.m_pImpl;
+}
+
+int GroupHandle::index() const {
+    if (!m_pImpl) {
+        return kInvalidIndex;
+    }
+    DEBUG_ASSERT(m_pImpl->valid());
+    return m_pImpl->m_index;
+}
+
+QString GroupHandle::name() const {
+    if (!m_pImpl) {
+        return {};
+    }
+    DEBUG_ASSERT(m_pImpl->valid());
+    return m_pImpl->m_name;
+}
```

Comment:
why did you move this out the the header? They're trivial enough for that IMO.

Reply:
The definition of `m_pImpl` is not available in the header.

Reply:
Mhmm. can't we put it in the header in an anonymous namespace?

Reply:
Anonymous namespaces have no effect. That's not how the [pImpl](https://en.cppreference.com/w/cpp/language/pimpl) idiom works.

Reply:
Performance is not an issue here.

Reply:
Mhmm. But why use the pimpl idiom at all here then? We don't have to provide a stable ABI and we usually don't really care about compile-times either.

Reply:
That's the whole point of this PR. Handles are shallow wrappers around a plain pointer and could be passed/copied/cloned by std::memcpy -> Q_MOVABLE_TYPE. Using the plain pointer without the wrapper is unhandy and leaks implementation details.

Reply:
Can you elaborate more on why the pointer would leak implementation details? Sorry for the dumb question but I struggle to understand the actual benefit a handle like this has over a pointer or a reference. The introduced type doesn't help with lifetime/ownership so I struggle to see the purpose.

Reply:
The typedef to the plain pointer caused clazy warnings with `auto` that you could not resolve without this wrapper.

Reply:
leaking = "the handle is actually a plain pointer"

Reply:
Thats not really what I meant. I thought previously, the channelhandle was an index and a name. I agree that there is a benefit of bundling those up into a class (lets call it `A`). But why do we then need to create another wrapper class `B` (the new channel handle) that wraps `A` (channelhandleimpl)?

Reply:
Because you cannot pass ChannelHandleImpl by std::memcpy. It is not a POD.

Reply:
Right, but a `*ChannelHandleImpl` pointer is trivial and thus memcpyable. So what is the advantage of the newtype over the plain pointer?

Reply:
As I already explained you get compiler warnings if you try to use it like `auto groupHandle = ...`. Using `auto* groupHandle = ...` as suggested reveals the implementation details, which is not desired.

Reply:
I wasn't aware of this wrapper. If it works well with Qt you could try to replace it. Since the pointer is immutable storing a reference instead is suitable.

Reply:
Though `std::reference_wrapper<GroupHandleImpl>` might be dangerous, because it doesn't handle `nullptr`.

Reply:
Well, it would be trivial to add a `using GroupHandle = std::reference_wrapper<const GroupHandleImpl>`. Can you elaborate what you mean in regards to the safety? You can't construct a `std::reference_wrapper<const GroupHandleImpl>` from a pointer, only from a reference or the real type. If you want to construct it from a pointer, you have to manually deref the pointer, and thus ur responsible if its `nullptr`. 

I think the bigger problem is that in order to be compatible with Qt, `std::reference_wrapper` would have to be default constructible. I guess that would technically be possible if the underlying type was default constructible, but that feature is not in the standard unfortunately.

Reply:
I'd still like to use std::reference_wrapper instead of the newtype because it saves maintenance overhead.

Reply:
We need to wrap a nullptr for the default instance. That doesn't seem to be possible wit std::reference_wrapper, which also doesn't provide an `operator bool`.

Reply:
Well. IMO conceptually there should not be a default instance of a grouphandle IMO. If we really need one, we could make it known global static object instead of `nullptr`.